src/adaptors/adaptor_tls.c

```diff
@@ -723,7 +729,289 @@ bool qd_tls_has_output(const qd_tls_t *tls)
     return tls->tls_has_output;
 }
 
+bool qd_tls_is_input_drained(const qd_tls_t *tls, bool *close_notify)
+{
+    assert(tls);
+    *close_notify = pn_tls_is_input_closed(tls->tls_session);
+    return tls->input_drained;
+}
+
+bool qd_tls_is_output_flushed(const qd_tls_t *tls)
+{
+    assert(tls);
+    return tls->output_flushed;
+}
+
 pn_tls_t *qd_tls_get_pn_tls_session(qd_tls_t *tls)
 {
+    assert(tls);
     return tls->tls_session;
 }
+
+// note: only use with qd_tls_do_io
+uint64_t qd_tls_get_encrypted_output_octet_count(const qd_tls_t *tls)
+{
+    assert(tls);
+    return tls->encrypted_output_bytes;
+}
+
+// note: only use with qd_tls_do_io
+uint64_t qd_tls_get_encrypted_input_octet_count(const qd_tls_t *tls)
+{
+    assert(tls);
+    return tls->encrypted_input_bytes;
+}
+
+int qd_tls_do_io(qd_tls_t                     *tls,
+                 pn_raw_connection_t          *raw_conn,
+                 qd_tls_take_output_data_cb_t *take_output_cb,
+                 void                         *take_output_context,
+                 qd_adaptor_buffer_list_t     *input_data,
+                 uint64_t                     *input_data_count)
+{
+    bool work;
+    *input_data_count = 0;
+
+    assert(tls);
+    assert(raw_conn);
+
+    do {
+        size_t          capacity;
+        size_t          taken;
+        size_t          given;
+        pn_raw_buffer_t pn_buf_desc;
+        uint64_t        total_octets;
+
+        // Loop until no more work can be done. "work" is considered true whenever the TLS layer produces output or
+        // opens up capacity for more input
+
+        work = false;
+
+        // give empty buffers for holding encrypted output from TLS
+
+        capacity = pn_tls_get_encrypt_output_buffer_capacity(tls->tls_session);
+        while (capacity-- > 0) {
+            (void) qd_adaptor_buffer_raw(&pn_buf_desc);
+            given = pn_tls_give_encrypt_output_buffers(tls->tls_session, &pn_buf_desc, 1);
+            (void) given;
+            assert(given == 1);
+        }
+
+        // push any unencrypted output data from the protocol adaptor into TLS for encryption.
+
+        if (pn_tls_is_secure(tls->tls_session)) {
+            assert(take_output_cb);
+            capacity = pn_tls_get_encrypt_input_buffer_capacity(tls->tls_session);
+
+            if (capacity > 0) {
+                qd_adaptor_buffer_list_t ubufs = DEQ_EMPTY;
+
+                int64_t out_octets = take_output_cb(take_output_context, &ubufs, capacity);
```

Comment:
octets are not necessary.

Reply:
Out of date: out_octets are used in the trace log output. Very helpful for debug.